lib/drivers/abbott/freeStyleLibreProtocol.js

```diff
@@ -0,0 +1,665 @@
+
+const DB_RECORD_NUMBER_REGEX = new RegExp('^DB Record Number = ([0-9]+)$');
+
+// after db record responses (e.g. $arresult?, $history?) an additional line is send,
+// containing the record count and an additional checksum of just the records:
+// COUNT,RECORD_CHECKSUM\r\n
+// eslint-disable-next-line no-control-regex
+const DB_RECORDS_REGEX = new RegExp('^([^]*\r\n)([0-9]+),([0-9A-F]{8})$');
+
+
+export default class FreeStyleLibreProtocol {
+  constructor(config) {
+    this.config = config;
+    this.hidDevice = config.deviceComms;
+    this.sequenceRx = 0;
+    this.sequenceTx = 0;
+    this.lastAckRx = 0;
+  }
+
+  readResponse(responseType, timeout, cb) {
+    let receiveTimeout = null;
+    const resetReceiveTimeout = () => {
+      if (receiveTimeout !== null) {
+        clearTimeout(receiveTimeout);
+      }
+      receiveTimeout = setTimeout(() => {
+        debug('readResponse: TIMEOUT');
+        const e = new Error('Timeout error.');
+        e.name = 'TIMEOUT';
+        return cb(e, null);
+      }, timeout);
+    };
+
+    let ackInterval = null;
+    if (responseType === COMMAND.BINARY_RESPONSE) {
+      ackInterval = setInterval(() => {
+        if (this.sequenceRx !== this.lastAckRx) {
+          this.sendAck();
+        }
+      }, ACK_INTERVAL);
+    }
+
+    const aapPackets = [];
+    let receiveBuffer = new Buffer(0);
+    async.doWhilst(
+      (callback) => {
+        resetReceiveTimeout();
+        this.hidDevice.receive((buffer) => {
+          if (buffer !== undefined && buffer.length > 0) {
+            const hidPacket = this.constructor.parseHidPacket(buffer);
+
+            if (hidPacket.responseType === COMMAND.TEXT_RESPONSE) {
+              // append newly received data to message string
+              receiveBuffer = Buffer.concat([receiveBuffer, hidPacket.data]);
+            } else if (hidPacket.responseType === COMMAND.BINARY_RESPONSE) {
+              const atpPacket = this.constructor.parseAtpFrame(hidPacket.data);
+              if (atpPacket === null) {
+                // CRC error: try to get damaged package again
+                this.sendAck();
+              } else if (this.sequenceRx === atpPacket.sequenceTx) {
+                // next sequence number expected is one higher than the one just received
+                this.sequenceRx = (this.sequenceRx + 1) % 0x100;
+
+                receiveBuffer = Buffer.concat([receiveBuffer, atpPacket.data]);
+              } else {
+                debug('readResponse: Received wrong sequence number', atpPacket.sequenceTx,
+                  ', expected', this.sequenceRx);
+                // try to get missing package by sending an ACK containing the next expected
+                // sequence number to the device
+                this.sendAck();
+              }
+            } else if (hidPacket.responseType === COMMAND.ACK_FROM_DEVICE) {
+
+              // ignore these for now
```

Comment:
Should this be implemented? If not, can you please update the comment with what the expected behaviour should be when this case is not ignored, or why it's ignored (similarly to https://github.com/tidepool-org/chrome-uploader/pull/469/files#diff-9f2e95903da58a2a654b636f764330d6R202)?